raysort/main_push.py

```diff
@@ -185,34 +186,113 @@ def reduce(self) -> list[PartInfo]:
             for r, merge_out in enumerate(merge_results):
                 if len(tasks_in_flight) >= self.cfg.reduce_parallelism:
                     _, tasks_in_flight = ray.wait(tasks_in_flight, fetch_local=False)
+
                 ref = final_merge.options(**ray_utils.current_node_aff()).remote(
-                    self.cfg, self.worker_id, r, *merge_out
+                    self.cfg, self.worker_id, r, list(merge_out)
                 )
                 merge_results[r, :] = None
-                tasks_in_flight.append(ref)
-                results.append(ref)
+                if isinstance(ref, list):
+                    tasks_in_flight.extend(ref)
+                    results.extend(ref)
+                else:
+                    tasks_in_flight.append(ref)
+                    results.append(ref)
             return [r for r in ray.get(results) if r is not None]
 
 
+def get_nonempty_part(parts: list[ray.ObjectRef]) -> Optional[np.ndarray]:
+    part_0 = ray.get(parts[0])
+    counter = 1
+    while len(part_0) == 0 and counter < len(parts):
+        part_0 = ray.get(parts[counter])
+        counter += 1
+    return part_0
+
+
+def partition_parts(
+    worker_id: PartId, reduce_idx: PartId, parts: list[ray.ObjectRef]
+) -> Tuple[list[ray.ObjectRef], list[ray.ObjectRef]]:
+    part = get_nonempty_part(parts)
+
+    pivot = sort_utils.get_median_key(part)
+    separated_parts = [
+        sort_utils.split_part.remote(
+            part, pivot, (str(worker_id) + "_" + str(reduce_idx))
+        )
+        for part in parts
+    ]
+
+    first_parts = [a for a, _ in separated_parts if a]
+    second_parts = [b for _, b in separated_parts if b]
+
+    return first_parts, second_parts
+
+
 # Memory usage: merge_partitions.batch_num_records * RECORD_SIZE = 100MB
 # Plasma usage: input_part_size = 2GB
 @ray.remote(num_cpus=0)
 def final_merge(
     cfg: AppConfig,
     worker_id: PartId,
     reduce_idx: PartId,
-    *parts: list[np.ndarray],
-) -> Optional[PartInfo]:
+    parts: list[np.ndarray],
+    subpart_idx: PartId = 1,
+    level: int = 1,
+) -> list[PartInfo]:
     logging_utils.init()
+
     with tracing_utils.timeit("reduce"):
         M = len(parts)
+        if M == 0:
+            return []
+
+        ray.wait(parts, num_returns=len(parts))
```

Comment:
We can also take this part out as a helper function `_get_total_partition_sizes()`. I'd like to break up long functions as much as I could.